qa: Clarify assert_start_raises_init_error output

hodlinator opened the pull request:
Raising a new exception from within a Python `except`-block, as `assert_start_raises_init_error()` does, causes the interpreter to generate extra error output which is unnecessary in this case.

<details><summary>Example output before & after this PR</summary>

Before:
```
2025-07-08T20:05:48.407001Z TestFramework (ERROR): Assertion failed
Traceback (most recent call last):
  File "/home/hodlinator/bitcoin/test/functional/test_framework/test_node.py", line 686, in assert_start_raises_init_error
    ret = self.process.wait(timeout=self.rpc_timeout)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/fqm9bqqlmaqqr02qbalm1bazp810qfiw-python3-3.12.9/lib/python3.12/subprocess.py", line 1266, in wait
    return self._wait(timeout=timeout)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/fqm9bqqlmaqqr02qbalm1bazp810qfiw-python3-3.12.9/lib/python3.12/subprocess.py", line 2053, in _wait
    raise TimeoutExpired(self.args, timeout)
subprocess.TimeoutExpired: Command '['/home/hodlinator/bitcoin/build/bin/bitcoind', '-datadir=/tmp/bitcoin_func_test_v96lkcq8/eb2665c7/node0', '-logtimemicros', '-debug', '-debugexclude=libevent', '-debugexclude=leveldb', '-debugexclude=rand', '-uacomment=testnode0', '-disablewallet', '-logthreadnames', '-logsourcelocations', '-loglevel=trace', '-v2transport=0']' timed out after 3 seconds

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/hodlinator/bitcoin/test/functional/test_framework/test_framework.py", line 186, in main
    self.setup()
  File "/home/hodlinator/bitcoin/test/functional/test_framework/test_framework.py", line 358, in setup
    self.setup_network()
  File "/home/hodlinator/bitcoin/build/test/functional/feature_framework_startup_failures.py", line 151, in setup_network
    self.nodes[0].assert_start_raises_init_error()
  File "/home/hodlinator/bitcoin/test/functional/test_framework/test_node.py", line 716, in assert_start_raises_init_error
    self._raise_assertion_error(assert_msg)
  File "/home/hodlinator/bitcoin/test/functional/test_framework/test_node.py", line 196, in _raise_assertion_error
    raise AssertionError(self._node_msg(msg))
AssertionError: [node 0] bitcoind should have exited within 3s with an error
```
After:
```
2025-07-08T20:09:15.330589Z TestFramework (ERROR): Assertion failed
Traceback (most recent call last):
  File "/home/hodlinator/bitcoin/test/functional/test_framework/test_framework.py", line 186, in main
    self.setup()
  File "/home/hodlinator/bitcoin/test/functional/test_framework/test_framework.py", line 358, in setup
    self.setup_network()
  File "/home/hodlinator/bitcoin/build/test/functional/feature_framework_startup_failures.py", line 151, in setup_network
    self.nodes[0].assert_start_raises_init_error()
  File "/home/hodlinator/bitcoin/test/functional/test_framework/test_node.py", line 720, in assert_start_raises_init_error
    self._raise_assertion_error(assert_msg)
  File "/home/hodlinator/bitcoin/test/functional/test_framework/test_node.py", line 196, in _raise_assertion_error
    raise AssertionError(self._node_msg(msg))
AssertionError: [node 0] bitcoind should have exited within 3s with an error (cmd: ['/home/hodlinator/bitcoin/build/bin/bitcoind', '-datadir=/tmp/bitcoin_func_test_v96lkcq8/eb2665c7/node0', '-logtimemicros', '-debug', '-debugexclude=libevent', '-debugexclude=leveldb', '-debugexclude=rand', '-uacomment=testnode0', '-disablewallet', '-logthreadnames', '-logsourcelocations', '-loglevel=trace', '-v2transport=0'])
```

</details>

---

Can be tested on this PR by:
1. Execute test containing new test case:
    ```shell
    build/test/functional/feature_framework_startup_failures.py -ldebug > after.log
    ```
2. Drop first commit which contains the fix.
3. Re-run test:
    ```shell
    build/test/functional/feature_framework_startup_failures.py -ldebug > before.log
    ```
4. Diff logs, focusing on `TestInitErrorTimeout OUTPUT` sections.

---

Found while testing #32835 using the suggested method (https://github.com/bitcoin/bitcoin/pull/32835#issue-3188748624) which triggered expected timeouts, but with the extra error noise.

hodlinator committed:
refactor(qa-tests): Extract InternalDurationTestMixin for use in next commit

hodlinator committed:
qa-tests: Log expected output in debug

Helpful when comparing expected/unexpected outputs against each other for working/broken code.

Also account for TimeoutExpired.output being None and halt instead of re-raising.

ryanofsky reviewed (approved):
Code review ACK fde4d2f0e1cf1252cbb24aad475332d96f3e6e80. Very welcome change! I almost always get confused by "During handling of the above exception, another exception occurred" messages, so it is great to eliminate one and stop showing a spurious internal exception before the real error when `assert_start_raises_init_error` checks fail.

Implementation of this PR is also a little simpler that it may first appear. The actual fix is small and is in the first commit, and the later commits are confined to the `feature_framework_startup_failures.py` test, ensuring the fix works, and making minor improvements there.

hodlinator committed:
fix(qa-tests): Bring back decoding of exception field

Partial revert of fab085c15f7221986f73af7e05e799edf3eadaf0

subprocess.run(..., text=True) is not sufficient to make the TimeoutExpired exception stop using byte buffers. This is a known issue, see: https://github.com/python/cpython/issues/87597

hodlinator commented:
Latest push (42ed679f0bb9807b08e6209528829f1e45f8765d):
* Rebases to resolve conflict with fab085c15f7221986f73af7e05e799edf3eadaf0
* Adds 9481948e0d979dccedd7b4e6616ef4903bc1cdba - Reverts one line of fab085c15f7221986f73af7e05e799edf3eadaf0 to fix issue
* Adds 0ff74253858f0a9b6c1f0fecb815c8380d5661be - Minor documentation change to `assert_start_raises_init_error()` as requested by https://github.com/bitcoin/bitcoin/pull/32929#issuecomment-3453237274
* Modifies last commit 42ed679f0bb9807b08e6209528829f1e45f8765d - Account for `e.output` being `None` and halt instead of re-raising exception.

Edit: Corrected reference to fab085c15f7221986f73af7e05e799edf3eadaf0.

furszy reviewed (commented):
Concept ACK, cool stuff

hodlinator committed:
qa: Improve assert_start_raises_init_error output

Re-raising within the except-block would trigger excessive "During handling of the above exception, another exception occurred"-output.

Also changed comment - exceptions are raised in Python, not thrown.

ryanofsky reviewed (commented):
Code review ACK 6b0eac750f3aecb29446d70c039559143e43a011. Since last review, just tweaking the last commit and adding two initial commits after some apparent conflicts with other PRs.

I do think this PR could benefit from being organized better and appearing to have more focus. I would definitely move the third commit d7ff1d0fa572ec36ff1eb10c0a1c04ef5e5be072 first since it's the main commit and probably the one most useful for other developers.

I also think it would be good to distinguish improvements to the qa framework from improvements to the feature_framework test. I might go with different prefixes in commit titles like:

- qa: Make assert_start_raises_init_error output more readable
- qa-tests: Bring back decoding of exception field
- qa-tests: refactor: Extract InternalDurationTestMixin for use in next commit
- qa-tests: Add test for missing init errors
- qa-tests: Log expected output in debug
- doc: Make comment more pythonic

hodlinator committed:
qa-tests: Add test for timeouts due to missing init errors

Verifies that the fix to assert_start_raises_init_error in an earlier commit stays intact, with only one exception being raised instead of multiple.

hodlinator reviewed (commented):
Latest push (6b0eac750f3aecb29446d70c039559143e43a011 -> 356883f0e48be59bcb154096cef82cbf3f0dd9d8) was largely motivated by https://github.com/bitcoin/bitcoin/pull/32929#pullrequestreview-3584856747 and contains changes:
* Moves the main fix commit first, and merges in the tiny doc-commit that was updating a comment in the same function.
* Renames `TestInitErrorTimeout` -> `TestMissingInitErrorTimeout` and clarifies log (https://github.com/bitcoin/bitcoin/pull/32929#discussion_r2624666464).
* Improves commit messages.

ryanofsky reviewed (approved):
Code review ACK 356883f0e48be59bcb154096cef82cbf3f0dd9d8. Thanks for the updates! Just rearranged commits and made minor changes in "missing init errors" test since last review

Might also suggest renaming PR to "qa: Make assert_start_raises_init_error output more readable" to motivate it a little more. Current title is fine but to me me at least it wasn't obvious what a knock-on exception what the benefit would be of avoiding it.

l0rinc commented:
Only reviewed the diff [since my last ack](https://github.com/bitcoin/bitcoin/compare/6b0eac750f3aecb29446d70c039559143e43a011..356883f0e48be59bcb154096cef82cbf3f0dd9d8) which only contains the rename and a description update - and some commit structure and message changes which I was already fine with before.

> Might also suggest renaming PR to "qa: Make assert_start_raises_init_error output more readable" 

The PR has a PR problem

ACK 356883f0e48be59bcb154096cef82cbf3f0dd9d8

furszy reviewed (commented):
Code ACK 356883f0e48be59bcb154096cef82cbf3f0dd9d8
